lib/modules/platform/scm-manager/index.spec.ts

```diff
@@ -0,0 +1,644 @@
+          pageTotal: 1,
+          _embedded: {
+            pullRequests: [],
+          },
+        });
+
+      expect(await getPrList()).toIncludeAllMembers([]);
+    });
+
+    it('should return empty array, because api request failed', async () => {
+      httpMock
+        .scope(endpoint)
+        .get(
+          `/pull-requests/${repo.namespace}/${repo.name}?status=ALL&pageSize=1000000`,
+        )
+        .reply(400);
+
+      expect(await getPrList()).toIncludeAllMembers([]);
+    });
+
+    it('should return all PRs of a repo', async () => {
+      const expectedResult: Pr[] = [
+        {
+          sourceBranch: pullRequest.source,
+          createdAt: pullRequest.creationDate,
+          labels: pullRequest.labels,
+          number: parseInt(pullRequest.id),
+          state: pullRequest.status,
+          targetBranch: pullRequest.target,
+          title: pullRequest.title,
+          hasAssignees: false,
+          isDraft: false,
+          reviewers: [],
+        },
+      ];
+
+      httpMock
+        .scope(endpoint)
+        .get(
+          `/pull-requests/${repo.namespace}/${repo.name}?status=ALL&pageSize=1000000`,
+        )
+        .reply(200, {
+          page: 0,
+          pageTotal: 1,
+          _embedded: {
+            pullRequests: [pullRequest],
+          },
+        });
+
+      //Fetching from client
+      expect(await getPrList()).toIncludeAllMembers(expectedResult);
+      //Fetching from cache
+      expect(await getPrList()).toIncludeAllMembers(expectedResult);
+    });
+  });
+
+  describe(findPr, () => {
+    it('search in pull request without explicitly setting the state as argument', async () => {
+      httpMock
+        .scope(endpoint)
+        .get(
+          `/pull-requests/${repo.namespace}/${repo.name}?status=ALL&pageSize=1000000`,
+        )
+        .reply(200, {
+          page: 0,
+          pageTotal: 1,
+          _embedded: {
+            pullRequests: [pullRequest],
+          },
+        });
+
+      expect(
+        await findPr({
+          branchName: pullRequest.source,
+          prTitle: pullRequest.title,
+        }),
+      ).toEqual(renovatePr);
+    });
+
+    it.each([
+      [[], pullRequest.source, pullRequest.title, 'all', null],
+      [[pullRequest], 'invalid branchName', pullRequest.title, 'all', null],
+      [[pullRequest], pullRequest.source, 'invalid title', 'all', null],
+      [[pullRequest], pullRequest.source, null, 'all', renovatePr],
+      [[pullRequest], pullRequest.source, undefined, 'all', renovatePr],
+      [[pullRequest], pullRequest.source, pullRequest.title, 'all', renovatePr],
+      [
+        [pullRequest],
+        pullRequest.source,
+        pullRequest.title,
+        'open',
+        renovatePr,
+      ],
+      [[pullRequest], pullRequest.source, pullRequest.title, '!open', null],
+      [[pullRequest], pullRequest.source, pullRequest.title, 'closed', null],
+    ])(
+      'search within %p for %p, %p, %p with result %p',
```

Comment:
This code looks hard to understand and maintain.

I don't know what `'search within %p for %p, %p, %p with result %p',` means here.

Maybe a maintainer can suggest improvements?

Reply:
I find this tabular approach is fairly readable: https://github.com/renovatebot/renovate/blob/85267b9861cffc680a5efc4ae8266791232bffa7/lib/modules/versioning/poetry/index.spec.ts#L28-L45